### ISSUE TITLE
LearningRateControl should be able to use combined losses

### ISSUE THREAD

**Issue by ChristopherBrix** [rwth-i6/returnn#195]
I just realized that the learning rate control logic depends on the name of the output layer, if multiple outputs are defined.

If there is only one output defined, [`dev_score` and `dev_error` are set to the respective values of the output layer](https://github.com/rwth-i6/returnn/blob/master/LearningRateControl.py#L260).
For multiple outputs, [they are stored like `dev_score_LAYER_NAME`](https://github.com/rwth-i6/returnn/blob/master/LearningRateControl.py#L262).

Later, these values are needed to determine whether the LR needs to be updated. For one output, [this is easy, because the key value (`dev_score`) is known](https://github.com/rwth-i6/returnn/blob/master/LearningRateControl.py#L296).
For multiple outputs, where no output layer is called `output`, [the first (alphabetically sorted) `dev_` entry is used](https://github.com/rwth-i6/returnn/blob/master/LearningRateControl.py#L306).

This is already not intuitive, I would have expected that for two outputs, the two losses are eg. added.
But in my case, the `decision` layer happens to be the first layer found. But its value is 0, so later,  [the computation of the relative error leads to NaN](https://github.com/rwth-i6/returnn/blob/master/LearningRateControl.py#L235), and my learning rate is never reduced.

Before I write the PR for this, I would like to clarify whether there is a reason to choose just one of the output layers for the LR scheduling. Should the default be to add all losses, and do we need a flag to ensure backwards compatibility?

**Comment by JackTemaki** [comment]
Do I understand it correct, that the return value (key) is arbitrary (alphabetic) when there is no layer called 'output'?

**Comment by ChristopherBrix** [comment]
If you only have one output layer, then the name should not matter. 
But if you have multiple output layers, then yes, the alphabetically first output layer name is chosen for the learning rate scheduling.

**Comment by ChristopherBrix** [comment]
Oh, but you probably have a `decision` layer, right? And for all of my experiments, this layer logs an error of 0. This never really bothered me, I just ignored it. But if you have the same setup, then you automatically have at least 2 output layers, and now the name of your "real" output layer does matter. It should be "output", or some alphabetically smaller name than "decision". 
I will test this later. It should be visible in the logs, if your learning rate never decreases.

**Comment by JackTemaki** [comment]
This is my learning rate file with multiple outputs (no decision though):

```
200: EpochData(learningRate=6.461081889226677e-05, error={
'dev_error_decoder/output': 0.7288903655845147,
'dev_error_decoder/stop_token': 0.0,
'dev_error_mse_output': 0.6737158583869719,
'dev_error_output': 0.6737158583869719,
'dev_score_decoder/entropy': 0.7084714989882942,
'dev_score_decoder/output': 0.3153343942486333,
'dev_score_decoder/stop_token': 0.01600229209713947,
'dev_score_mse_output': 0.1126947590044958,
'dev_score_output': 0.25680778074465826,
'train_error_decoder/output': 0.7102662612267641,
'train_error_decoder/stop_token': 0.0,
'train_error_mse_output': 0.6731171007399964,
'train_error_output': 0.6731171007399964,
'train_score_decoder/entropy': 0.7582645078670691,
'train_score_decoder/output': 0.29592065600534634,
'train_score_decoder/stop_token': 0.014067114256025644,
'train_score_mse_output': 0.10915044330076043,
'train_score_output': 0.2521381113616365,
}),
```
but as the learning rate is reduced correctly, I always assumed that dev_score_output is the only number used.

**Comment by ChristopherBrix** [comment]
So you have multiple output layers, and one of them is called "output", right? Then this layer is responsible for the score you see reported for 'dev_score_output'. And only this value will be used for the learning rate scheduling.

But if you were to call this layer "my_output", then the alphabetically smallest layer name would be used.

If this is the intended behavior, it answers my first question: The losses should not all be added. So we need some kind of flag for this. And I'd say that there should be a warning if no output layer is called "output".

**Comment by JackTemaki** [comment]
Yes, this is definitely not as it is supposed to be. My suggestion would be (in addition to the warning): a flag for each layer that should be used to control the learning rate, and all losses that have the flag are added. In most cases one should be sufficient, but there are cases where you want more (for example, to control it by L1 and L2 loss together)

**Comment by albertz** [comment]
The current behavior is exactly as intended.

But maybe you want to have it more flexible by summing or even something
more complex. Maybe you can just specify an eval string (or callable
function).


Nick Rossenbach <notifications@github.com> schrieb am Fr., 5. Juli 2019,
16:18:

**Comment by JackTemaki** [comment]
What is the reason for alphabetic ordering? I do not understand how that can be correct in any case...

**Comment by albertz** [comment]
It needs to pick one. So if you did not configure it, it should at least be
deterministic, so that was the most canonical thing to do.

Nick Rossenbach <notifications@github.com> schrieb am Fr., 5. Juli 2019,
16:27:

**Comment by JackTemaki** [comment]
But when I understood Christoph correctly, you can not configure it unless you change your layer names, and I think it should either crash or at least give a noticeable warning in that case, as I do not see a case where this would lead to the desired training procedure.

**Comment by albertz** [comment]
You can configure it already.

Nick Rossenbach <notifications@github.com> schrieb am Fr., 5. Juli 2019,
16:36:

**Comment by ChristopherBrix** [comment]
I've implemented an option to define an arbitrary `MetaLoss` which will be computed for both training and scoring, and is logged in the same way as all other losses. See my PR: #196 

How would you currently configure which loss is used for the LR scheduling? I would have just added it as another special case to the [`get_error_key`](https://github.com/rwth-i6/returnn/blob/master/LearningRateControl.py#L272) function, but I don't want to duplicate anything.

**Comment by albertz** [comment]
Is this actually something which you really need right now, or is an single
error key enough (which you can already configure)?


Christopher Brix <notifications@github.com> schrieb am Fr., 5. Juli 2019,
23:05:

**Comment by JackTemaki** [comment]
From the code i did not see directly how to configure it, sorry. I just updated the docs accordingly. One more question remains then, what happens if you provide a list for `learning_rate_control_error_measure`? For me it seems like only the first entry is used.

**Comment by JackTemaki** [comment]
@ChristopherBrix this should also answer your question, if i understood the code correctly.
>How would you currently configure which loss is used for the LR scheduling? I would have just added 
> it as another special case to the get_error_key function, but I don't want to duplicate anything.

**Comment by albertz** [comment]
I think the first available entry is used, if it is a list (if I remember
correctly). Eg via pretraining or whatever, you might have different error
entries in each epoch.



Nick Rossenbach <notifications@github.com> schrieb am Sa., 6. Juli 2019,
10:43:

**Comment by ChristopherBrix** [comment]
> Is this actually something which you really need right now, or is an single error key enough (which you can already configure)?
Yes, I need this right now: I'm training one model to solve two similar problems (eg. translate to two target languages), and would like the learning rate to depend on the combined learning progress.

If you think this is not a common situation, I can keep this change in my private fork. Otherwise, I'll complete the PR so it can be merged.

**Comment by JackTemaki** [comment]
> If you think this is not a common situation [...]

I could definitely use this for my experiments as well, as i have multiple losses for multiple layers, so a pull request would be nice from my point of view ;)

**Comment by albertz** [comment]
I was just asking whether you will actually use it now, not whether it
might be useful.


Nick Rossenbach <notifications@github.com> schrieb am Sa., 6. Juli 2019,
14:44:

**Comment by JackTemaki** [comment]
> I was just asking whether you will actually use it now, not whether it might be useful.

I would directly use it if it is there, but it is not crucial for the experiments, as the losses are usually correlated enough.

**Comment by ChristopherBrix** [comment]
I've updated the PR. Now, you can create arbitrary many custom meta losses. If you want to use it for the learning rate scheduling, you have to indicate so using the `learning_rate_control_error_measure` flag.

I also added an assertion that the relative error is never `NaN`, which happens if you activate the learning rate scheduling but no loss can be found. I'd rather have my experiments crash, then the learning rate to be (silently) constant. If one would like to have a constant LR, this can be defined in the config.

**Comment by ChristopherBrix** [comment]
As explained by @albertz, custom losses can be implemented using the eval layer, so this is not really needed.
